analysis/cpp/src/make-histograms-loop.cpp

```diff
@@ -60,33 +60,58 @@ int main(int argc, char* argv[]) {
     tree->Add(input_file.c_str());
     std::cout << "Loaded TTree \"" << tree_name << "\" with " << tree->GetEntries() << " events" << std::endl;
 
-    // primary vertex related variables
+    /*
+     Here we declare local variables that will be used to hold the data
+     for each of the variables we wish to read from the input
+     TTree file. There must be one locally declared variable for
+     each of the variables in the TTree that we wish to read.
+     */
+
+    // reco-level primary vertex related variables
     float pv_xpos = 0;
     float pv_ypos = 0;
     float pv_zpos = 0;
 
-    // variables for mc particles
-    int* mcpdg = new int[5000000]; // ROOT is dumb so just allocate a ton of memory for these arrays
-    float* mc_px = new float[5000000];
-    float* mc_py = new float[5000000];
-    float* mc_xpos = 0;
-    int n_mc_particles_in_event;
+    // truth-level primary vertex proxies (we take the vertex positions from the truth muons from the muon gun as a proxy)
+    float truth_pv_xpos = 0;
+    float truth_pv_ypos = 0;
+    float truth_pv_zpos = 0;
+
+    // mc particle related variables
+    const size_t mc_particle_array_size = 5000000;
+    int* mcpdg = new int[mc_particle_array_size]; // pdg id
+    int* mcgst = new int[mc_particle_array_size]; // MC generator status (1 == generated particle, 0 = decay ??)
+    float* mc_px = new float[mc_particle_array_size]; // x momentum
+    float* mc_py = new float[mc_particle_array_size]; // y momentum
+    float* mc_vtx_xpos = new float[mc_particle_array_size]; // x coordinate of MC particle vertex
+    float* mc_vtx_ypos = new float[mc_particle_array_size]; // y coordinate of MC particle vertex
+    float* mc_vtx_zpos = new float[mc_particle_array_size]; // z coordinate of MC particle vertex
+    int n_mc_particles_in_event = 0; // the number of MC particles in the event
 
     // attach the local variables to those in the TTree
+     tree->SetBranchAddress("nmcp", &n_mc_particles_in_event);
     tree->SetBranchAddress("vtxxx", &pv_xpos);
-    tree->SetBranchAddress("mcvtx", &pv_xpos);
-    // tree->SetBranchAddress("vtyyy", &pv_ypos);
-    // tree->SetBranchAddress("vtzzz", &pv_zpos);
-    // tree->SetBranchAddress("mcpdg", mcpdg);
-    // tree->SetBranchAddress("nmcp", &n_mc_particles_in_event);
-    // tree->SetBranchAddress("mcmox", mc_px);
-    // tree->SetBranchAddress("mcmoy", mc_py);
+    tree->SetBranchAddress("mcvtx", mc_vtx_xpos);
+//tree->SetBranchAddress("nmcp", &n_mc_particles_in_event); // we need to load the "nmcp" varialbe so that we know how many MC particles there are in an event
+     //tree->SetBranchAddress("vtyyy", &pv_ypos);
+     //tree->SetBranchAddress("vtzzz", &pv_zpos);
+     tree->SetBranchAddress("mcpdg", mcpdg);
+     tree->SetBranchAddress("mcgst", mcgst);
```

Comment:
Since we want to use `mcpdg` and `mcgst` in our analysis as we loop through events (below), then we must do this `SetBranchAddress` call: this links the variables contained in the TTree to the variables that we have locally defined in our script (in the lines above).